clang/lib/AST/AbstractSet.cpp

```diff
@@ -0,0 +1,31 @@
+#include "clang/AST/AbstractSet.h"
+
+using namespace clang;
+
+std::set<AbstractSet *, AbstractSetComparer> AbstractSetManager::SortedAbstractSets;
+
+AbstractSet *AbstractSetManager::GetOrCreateAbstractSet(Expr *E, ASTContext &Ctx) {
+  // Create a canonical form for E.
+  PreorderAST P(Ctx, E);
+  P.Normalize();
+
+  // Search for an existing AbstractSet in SortedAbstractSets that contains E,
+  // i.e. an AbstractSet whose CanonicalForm is equal to E's canonical form.
+  AbstractSet *Test = new AbstractSet(P);
+  auto It = SortedAbstractSets.find(Test);
+  if (It != SortedAbstractSets.end()) {
+    delete Test;
+    return *It;
+  }
+
+  // If there is no existing AbstractSet that contains E, create a new
+  // AbstractSet that contains E.
+  AbstractSet *A = new AbstractSet(P);
+  A->SetRepresentative(E);
+  SortedAbstractSets.emplace(A);
+  return A;

```

Comment:
`GetOrCreateAbstractSet` is unnecessarily creating a `Test`  `AbstractSet` that may later be immediately deleted (causes a perf hit) - the code in this function could be made better. This is important because this function will be called multiple times, and there may be many expressions that yield the same canonical form (the canonical form is agnostic to different `DeclRefExprs` for the same `VarDecls`).

Therefore, it will be great if the following suggestion can be considered:
We define a `SortedCanonicalForms` set with type `std::set<PreorderAST *, PreorderASTComparer>` (`AbstractSetComparer` above will become  `PreorderASTComparer` and move to `PreorderAST.h` - this will also be useful in other situations that use `PreorderAST`).
In addition, we maintain a map from canonical forms to abstract sets in the `AbstractSetManager` class.
This will help better organize the code in `GetOrCreateAbstractSet`.